Add support for GitLab

gep13 opened the pull request:
## Description

This PR adds support to allow creation, and maintenance of, releases on GitLab.  This has been made possible using the NGItLab library.  There are some items that have been left unimplemented, see the commit messages for more details.

## Related Issue

Fixes #146 

## Motivation and Context

I work on some projects that use GitLab for private work, and I want the ability to create release notes in the same way as done on public projects on GitHub.

## How Has This Been Tested?

1. Create both a public and private repository on GitLab
2. Get token to allow maintenance of these projects
3. Create a milestone and some issues using standard GRM labels
4. Run the `create` command to create release from the milestone - this will create an "upcoming release"
5. Run the `discard` command to delete the release
6. Run the `create` command again
7. Run the `publish` command to update the release to actually being releases
8. Run the `close` command to close the milestone
9. Run the `open` command to open the milestone
10. Run the `export` command to grab a markdown file version of the release that was created

The above are some sample things to do to verify things are working as expected, but various other tests have also been completed.  Let me know if you run into any problems with anything.

## Screenshots (if appropriate):

N#A

## Checklist:

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

AdmiringWorm commented on src/GitReleaseManager.Core/Provider/GitLabProvider.cs:
```diff
@@ -0,0 +1,435 @@
+                    };
+
+                    issueNotesClient.Create(issueComment);
+                }
```
```suggestion
                }

                return Task.CompletedTask;
```

AdmiringWorm commented on src/GitReleaseManager.Core/Provider/GitLabProvider.cs:
```diff
@@ -0,0 +1,435 @@
+
+        public Task<IEnumerable<Issue>> GetIssuesAsync(string owner, string repository, Milestone milestone, ItemStateFilter itemStateFilter = ItemStateFilter.All)
+        {
+            return ExecuteAsync(async () =>
```
```suggestion
            return ExecuteAsync(() =>
```

AdmiringWorm commented on src/GitReleaseManager.Core/Provider/GitLabProvider.cs:
```diff
@@ -0,0 +1,435 @@
+                issuesAndMergeRequests.AddRange(_mapper.Map<IEnumerable<Issue>>(issues));
+                issuesAndMergeRequests.AddRange(_mapper.Map<IEnumerable<Issue>>(mergeRequests));
+
+                return issuesAndMergeRequests.AsEnumerable();
```
```suggestion
                return Task.FromResult(issuesAndMergeRequests.AsEnumerable());
```

May not work. It may be that you need to use the explicit return result (not entirely sure).

AdmiringWorm commented on src/GitReleaseManager.Core/Provider/GitLabProvider.cs:
```diff
@@ -0,0 +1,435 @@
+
+        public Task<IEnumerable<IssueComment>> GetIssueCommentsAsync(string owner, string repository, Issue issue)
+        {
+            return ExecuteAsync(async () =>
```
```suggestion
            return ExecuteAsync(() =>
```

AdmiringWorm commented on src/GitReleaseManager.Core/Provider/GitLabProvider.cs:
```diff
@@ -0,0 +1,435 @@
+                    issueComments = _mapper.Map<IEnumerable<IssueComment>>(comments);
+                }
+
+                return issueComments;
```
```suggestion
                return Task.FromResult(issueComments);
```

AdmiringWorm commented on src/GitReleaseManager.Core/Provider/GitLabProvider.cs:
```diff
@@ -0,0 +1,435 @@
+
+        public Task<IEnumerable<Milestone>> GetMilestonesAsync(string owner, string repository, ItemStateFilter itemStateFilter = ItemStateFilter.All)
+        {
+            return ExecuteAsync(async () =>
```
```suggestion
            return ExecuteAsync(() =>
```

AdmiringWorm commented on src/GitReleaseManager.Core/Provider/GitLabProvider.cs:
```diff
@@ -0,0 +1,435 @@
+
+                foreach (var mappedMilestone in mappedMilestones)
+                {
+                    mappedMilestone.HtmlUrl = string.Format(CultureInfo.InvariantCulture, "https://gitlab.com/{0}/{1}/-/milestones/{2}#tab-issues", owner, repository, mappedMilestone.PublicNumber);
```
Can we extract this magic string format to a constant at the top of the file (or maybe into a different static class)?

AdmiringWorm commented on src/GitReleaseManager.Core/Provider/GitLabProvider.cs:
```diff
@@ -0,0 +1,435 @@
+                    mappedMilestone.HtmlUrl = string.Format(CultureInfo.InvariantCulture, "https://gitlab.com/{0}/{1}/-/milestones/{2}#tab-issues", owner, repository, mappedMilestone.PublicNumber);
+                }
+
+                return mappedMilestones;
```
```suggestion
                return Task.FromResult(mappedMilestones);
```

AdmiringWorm commented on src/GitReleaseManager.Core/Provider/GitLabProvider.cs:
```diff
@@ -0,0 +1,435 @@
+
+        public Task SetMilestoneStateAsync(string owner, string repository, Milestone milestone, ItemState itemState)
+        {
+            return ExecuteAsync(async () =>
```
```suggestion
            return ExecuteAsync(() =>
```

AdmiringWorm commented on src/GitReleaseManager.Core/Provider/GitLabProvider.cs:
```diff
@@ -0,0 +1,435 @@
+                else if (itemState == ItemState.Closed)
+                {
+                    mileStoneClient.Close(milestone.InternalNumber);
+                }
```
```suggestion
                }

                return Task.CompletedTask;
```

AdmiringWorm commented on src/GitReleaseManager.Core/Provider/GitLabProvider.cs:
```diff
@@ -0,0 +1,435 @@
+
+        public Task<Release> CreateReleaseAsync(string owner, string repository, Release release)
+        {
+            return ExecuteAsync(async () =>
```
```suggestion
            return ExecuteAsync(() =>
```

AdmiringWorm commented on src/GitReleaseManager.Core/Provider/GitLabProvider.cs:
```diff
@@ -0,0 +1,435 @@
+
+                if (mappedRelease != null)
+                {
+                    mappedRelease.HtmlUrl = string.Format(CultureInfo.InvariantCulture, "https://gitlab.com/{0}/{1}/-/releases/{2}", owner, repository, release.TagName);
```
Same as previously, it's probably best to extract out the magic string.

AdmiringWorm commented on src/GitReleaseManager.Core/Provider/GitLabProvider.cs:
```diff
@@ -0,0 +1,435 @@
+                    mappedRelease.HtmlUrl = string.Format(CultureInfo.InvariantCulture, "https://gitlab.com/{0}/{1}/-/releases/{2}", owner, repository, release.TagName);
+                }
+
+                return mappedRelease;
```
```suggestion
                return Task.FromResult(mappedRelease);
```

AdmiringWorm commented on src/GitReleaseManager.Core/Provider/GitLabProvider.cs:
```diff
@@ -0,0 +1,435 @@
+
+        public Task DeleteReleaseAsync(string owner, string repository, Release release)
+        {
+            return ExecuteAsync(async () =>
+            {
+                var releaseClient = _gitLabClient.GetReleases(GetGitLabProjectId(owner, repository));
+
+                releaseClient.Delete(release.TagName);
```
```suggestion
            return ExecuteAsync(() =>
            {
                var releaseClient = _gitLabClient.GetReleases(GetGitLabProjectId(owner, repository));

                releaseClient.Delete(release.TagName);
                return Task.CompletedTask;
```

AdmiringWorm commented on src/GitReleaseManager.Core/Provider/GitLabProvider.cs:
```diff
@@ -0,0 +1,435 @@
+
+        public Task PublishReleaseAsync(string owner, string repository, string tagName, Release release)
+        {
+            return ExecuteAsync(async () =>
```
```suggestion
            return ExecuteAsync(() =>
```

AdmiringWorm commented on src/GitReleaseManager.Core/Provider/GitLabProvider.cs:
```diff
@@ -0,0 +1,435 @@
+                    TagName = tagName,
+                };
+
+                releaseClient.Update(update);
```
```suggestion
                releaseClient.Update(update);
                return Task.CompletedTask;
```

AdmiringWorm commented on src/GitReleaseManager.Core/Provider/GitLabProvider.cs:
```diff
@@ -0,0 +1,435 @@
+
+        public Task UpdateReleaseAsync(string owner, string repository, Release release)
+        {
+            return ExecuteAsync(async () =>
```
```suggestion
            return ExecuteAsync(() =>
```

AdmiringWorm commented on src/GitReleaseManager.Core/Provider/GitLabProvider.cs:
```diff
@@ -0,0 +1,435 @@
+                    Milestones = new string[] { release.TagName },
+                };
+
+                releaseClient.Update(update);
```
```suggestion
                releaseClient.Update(update);
                return Task.CompletedTask;
```

AdmiringWorm commented on src/GitReleaseManager.Core/Provider/GitLabProvider.cs:
```diff
@@ -0,0 +1,435 @@
+
+        public string GetMilestoneQueryString()
+        {
+            return "sort=due_date_desc&state=closed";
```
Probably don't need the `sort=due_date_desc` at all.

AdmiringWorm commented on src/GitReleaseManager.Core/Provider/GitLabProvider.cs:
```diff
@@ -0,0 +1,435 @@
+            ////{
+            ////    throw new NotFoundException(ex.Message, ex);
+            ////}
+            catch (Exception ex) when (!(ex is NotFoundException))
```
We should probably also catch an [AggregateException](https://learn.microsoft.com/en-us/dotnet/api/system.aggregateexception?view=netstandard-2.0) which is commonly thrown in async functions.

AdmiringWorm commented on src/GitReleaseManager.Core/Provider/GitLabProvider.cs:
```diff
@@ -0,0 +1,435 @@
+            ////{
+            ////    throw new NotFoundException(ex.Message, ex);
+            ////}
+            catch (Exception ex) when (!(ex is NotFoundException))
```
Same as above.

AdmiringWorm commented on src/GitReleaseManager.Core/ReleaseNotes/ReleaseNotesBuilder.cs:
```diff
@@ -60,14 +60,16 @@ public async Task<string> BuildReleaseNotesAsync(string user, string repository,
             if (issues.Count == 0)
             {
-                var logMessage = string.Format("No closed issues have been found for milestone {0}, or all assigned issues are meant to be excluded from release notes, aborting creation of release.", _milestoneTitle);
+                var logMessage = string.Format(CultureInfo.InvariantCulture, "No closed issues have been found for milestone {0}, or all assigned issues are meant to be excluded from release notes, aborting creation of release.", _milestoneTitle);
```
Is this outputted to the user, intentionally?
If so this should use `CultureInfo.CurrentUiCulture`, not the invariant culture.

```suggestion
                var logMessage = string.Format(CultureInfo.InvariantCulture, "No closed issues have been found for milestone {0}, or all assigned issues are meant to be excluded from release notes, aborting release creation.", _milestoneTitle);
```

AdmiringWorm commented on src/GitReleaseManager.Core/VcsService.cs:
```diff
@@ -141,7 +142,7 @@ private async Task AddAssetsAsync(string owner, string repository, string tagNam
                         if (!File.Exists(asset))
                         {
-                            var message = string.Format("Requested asset to be uploaded doesn't exist: {0}", asset);
+                            var message = string.Format(CultureInfo.InvariantCulture, "Requested asset to be uploaded doesn't exist: {0}", asset);
```
Same as above, if this is meant to be outputted to the user eventually, it should be `CultureInfo.CurrentUiCulture`.

```suggestion
                            var message = string.Format(CultureInfo.InvariantCulture, "The requested asset to be uploaded doesn't exist: {0}", asset);
```

AdmiringWorm commented on src/GitReleaseManager.Core/VcsService.cs:
```diff
@@ -151,7 +152,15 @@ private async Task AddAssetsAsync(string owner, string repository, string tagNam
+
+                            if (_vcsProvider is GitLabProvider)
+                            {
+                                _logger.Error("Deleting of assets is not currently supported when targetting GitLab.");
```
```suggestion
                                _logger.Error("Deleting assets is not currently supported when targeting GitLab.");
```

AdmiringWorm commented on src/GitReleaseManager.Core/VcsService.cs:
```diff
@@ -177,7 +186,7 @@ private async Task AddAssetsAsync(string owner, string repository, string tagNam
                         {
                             _logger.Debug("Creating SHA section header");
-                            stringBuilder.AppendLine(string.Format("### {0}", _configuration.Create.ShaSectionHeading));
+                            stringBuilder.AppendLine(string.Format(CultureInfo.InvariantCulture, "### {0}", _configuration.Create.ShaSectionHeading));
```
```suggestion
                            stringBuilder.AppendFormat(CultureInfo.InvariantCulture, "### {0}", _configuration.Create.ShaSectionHeading).AppendLine();
```

or separate the calls into two calls

```suggestion
                            stringBuilder.AppendFormat(CultureInfo.InvariantCulture, "### {0}", _configuration.Create.ShaSectionHeading);
                            stringBuilder.AppendLine();
```

AdmiringWorm commented on src/GitReleaseManager.Core/VcsService.cs:
```diff
@@ -374,37 +383,38 @@ private async Task AddIssueCommentsAsync(string owner, string repository, Milest
                 catch (ForbiddenException)
                 {
-                    _logger.Error("Unable to add comment to issue #{IssueNumber}. Insufficient permissions.", issue.Number);
+                    _logger.Error("Unable to add comment to issue #{IssueNumber}. Insufficient permissions.", issue.PublicNumber);
```
```suggestion
                    _logger.Error("Unable to add a comment to issue #{IssueNumber}. Insufficient permissions.", issue.PublicNumber);
```

AdmiringWorm reviewed (changes requested):
In addition, is there any reason why the tests for GitHub specific tests aren't replicated for the GitLab provider?

gep13 commented on src/GitReleaseManager.Core/Provider/GitLabProvider.cs:
```diff
@@ -0,0 +1,435 @@
+            {
+                var releaseClient = _gitLabClient.GetReleases(GetGitLabProjectId(owner, repository));
+
+                return _mapper.Map<IEnumerable<Release>>(releaseClient.GetAsync());
```
Unfortunately not, no.  The method is:

```
GitLabCollectionResponse<ReleaseInfo> GetAsync(ReleaseQuery query = null);
```

gep13 commented on src/GitReleaseManager.Core/Options/BaseVcsSubOptions.cs:
```diff
@@ -14,5 +15,8 @@ public abstract class BaseVcsOptions : BaseSubOptions
         [Option('r', "repository", HelpText = "The name of the repository.", Required = true)]
         public string RepositoryName { get; set; }
+
+        [Option("provider", HelpText = "Version Control System provider", Default = VcsProvider.GitHub)]
```
Not in this initial release, no.  At some point, I think auto detecting would be a good addition, but outwith the scope of this PR.

gep13 commented on src/GitReleaseManager.Core/MappingProfiles/GitLabProfile.cs:
```diff
@@ -0,0 +1,48 @@
+                .ForMember(dest => dest.Description, act => act.MapFrom(src => src.Body))
+                .ForMember(dest => dest.Ref, act => act.MapFrom(src => src.TargetCommitish))
+                .ForMember(dest => dest.Milestones, act => act.MapFrom(src => new string[] { src.TagName }))
+                .ForMember(dest => dest.ReleasedAt, act => act.MapFrom(src => src.Draft ? DateTime.UtcNow.AddYears(1) : DateTime.UtcNow))
```
It ought to be :-P

gep13 commented on src/GitReleaseManager.Core/Model/Issue.cs:
```diff
@@ -6,10 +6,14 @@ public sealed class Issue
         public string Title { get; set; }
 
-        public int Number { get; set; }
+        public int InternalNumber { get; set; }
+
+        public int PublicNumber { get; set; }
```
Yes, I am ok with that.

gep13 commented on src/GitReleaseManager.Core/Model/Milestone.cs:
```diff
@@ -8,7 +8,9 @@ public sealed class Milestone
         public string Description { get; set; }
 
-        public int Number { get; set; }
+        public int PublicNumber { get; set; }
+
+        public int InternalNumber { get; set; }
```
Yip

gep13 commented on src/GitReleaseManager.Core/Extensions/MilestoneExtensions.cs:
```diff
@@ -8,7 +7,30 @@ public static class MilestoneExtensions
+                throw new ArgumentNullException(nameof(ver));
+            }
+
+            var nameWithoutPrerelease = ver.Title.Split('-')[0];
```
No, I believe that this is a requirement through the UI/API to provide a Title.

gep13 commented on src/GitReleaseManager.Cli/Program.cs:
```diff
@@ -197,5 +196,25 @@ private static Task<int> ExecuteCommand<TOptions>(TOptions options)
+            Log.Information("Using {Provider} as VCS Provider", vcsOptions.Provider);
+            if (vcsOptions.Provider == VcsProvider.GitLab)
+            {
+                var gitlabClient = new GitLabClient("https://gitlab.com", vcsOptions.Token);
```
As discussed, this needs to be:
```
.AddSingleton<IGitLabClient>((_) => new GitLabClient("https://gitlab.com", vcsOptions.Token))
```
But it has been added.

gep13 commented:
@AdmiringWorm if/when you get a minute, can you please have another review of this?  Thanks!

AdmiringWorm commented on src/GitReleaseManager.Core/Provider/GitLabProvider.cs:
```diff
@@ -0,0 +1,435 @@
+            Ensure.IsNotNullOrWhiteSpace(head, nameof(head));
+
+            return string.IsNullOrWhiteSpace(@base)
+                ? string.Format(CultureInfo.InvariantCulture, "https://gitlab.com/{0}/{1}/-/commits/{2}", owner, repository, head)
+                : string.Format(CultureInfo.InvariantCulture, "https://gitlab.com/{0}/{1}/-/compare/{2}...{3}", owner, repository, @base, head);
```
@gep13 any reason these weren't extracted so they aren't magic strings?

gep13 replied:
I distinctly remember making this change, that is why I added the :+1:, the changes must have been lost somehow 🤷🏻

AdmiringWorm commented on src/GitReleaseManager.Core/Provider/GitLabProvider.cs:
```diff
@@ -0,0 +1,434 @@
+
+        public Task<IEnumerable<Release>> GetReleasesAsync(string owner, string repository, bool skipPrereleases)
+        {
+            return ExecuteAsync(async () =>
+            {
+                var releaseClient = _gitLabClient.GetReleases(GetGitLabProjectId(owner, repository));
+
+                return _mapper.Map<IEnumerable<Release>>(releaseClient.GetAsync());
```
```suggestion
            return ExecuteAsync() =>
            {
                var releaseClient = _gitLabClient.GetReleases(GetGitLabProjectId(owner, repository));

                var release = _mapper.Map<IEnumerable<Release>>(releaseClient.GetAsync());
                return Task.FromResult(result);
```

AdmiringWorm commented on src/GitReleaseManager.Core/Provider/GitLabProvider.cs:
```diff
@@ -0,0 +1,434 @@
+            {
+                await action().ConfigureAwait(false);
+            }
+            catch (AggregateException ae)
+            {
+                throw new ApiException(ae.Message, ae);
+            }
```
The actual exception we are interested in is the inner exception, but only if it isn't null.

```suggestion
            catch (AggregateException ae) when (ae.InnerException != null)
            {
                throw new ApiException(ae.InnerException.Message, ae.InnerException);
            }
```

AdmiringWorm commented on src/GitReleaseManager.Core/Provider/GitLabProvider.cs:
```diff
@@ -0,0 +1,434 @@
+            {
+                return await action().ConfigureAwait(false);
+            }
+            catch (AggregateException ae)
+            {
+                throw new ApiException(ae.Message, ae);
+            }
```
```suggestion
            catch (AggregateException ae) when (ae.InnerException != null)
            {
                throw new ApiException(ae.InnerException.Message, ae.InnerException);
            }
```

gep13 commented:
@AdmiringWorm I have fixed up the remaining issues, and squashed them into the PR Review commit.  Once you are happy, I can look to squash that commit into other commits, or move forward with it as a standalone commit.

AdmiringWorm reviewed (approved):
LGTM.

@gep13 I'll leave it up to you to squash the Review commit whatever commit you feel is appropriate.

gep13 commented:
@AdmiringWorm I have squashed the final commit, and created some others, but the code itself hasn't changed.  I am going to move forward with merging this once the CI build completes.
